contracts/roulette/lib.rs

```diff
@@ -0,0 +1,388 @@
+
+            Ok(())
+        }
+
+        /// Spin the wheel
+        ///
+        /// Will also distribute the winnings to the players and reset the state, starting a new round of bets
+        #[ink(message)]
+        pub fn spin(&mut self) -> Result<()> {
+            if !self.is_betting_period_over() {
+                return Err(RouletteError::BettingPeriodNotOver);
+            };
+
+            // generate a "random" number between 1 and 36
+            // NOTE: this is a very poor source of randomness, we should add rcf palet
+            let winning_number = (self.env().block_timestamp() % 36 + 1) as u8;
+
+            self.distribute_payouts(winning_number)?;
+            self.reset()?;
+
+            Self::emit_event(self.env(), Event::WheelSpin(WheelSpin { winning_number }));
+
+            Ok(())
+        }
+
+        /// calculates anbd transfers payouts to the winning bets
+        fn distribute_payouts(&self, winning_number: u8) -> Result<()> {
+            let data = &self.data.get().unwrap();
+            (0..data.next_bet_id).try_for_each(|id| -> Result<()> {
+                let bet = self.bets.get(id).expect("all bet's should be present");
+                let payout = calculate_payout(&bet, Some(winning_number));
+
+                if payout > 0 {
+                    self.env()
+                        .transfer(bet.player, payout)
+                        .map_err(|why| RouletteError::NativeTransferFailed(format!("{why:?}")))?;
+                }
+
+                Ok(())
+            })?;
+
+            Ok(())
+        }
+
+        /// Reset the state allowing for a new round of bets to be made
+        fn reset(&mut self) -> Result<()> {
+            let mut data = self.data.get().unwrap();
+
+            data.betting_period_start = self.env().block_number();
+            data.next_bet_id = 0;
+            self.data.set(&data);
+
+            self.bets = Mapping::new();
+            self.account_id_to_bet = Mapping::new();
+
+            Ok(())
+        }
+
+        fn ensure_house(&self, caller: AccountId) -> Result<()> {
+            if self.data.get().unwrap().house.eq(&caller) {
+                Ok(())
+            } else {
+                Err(RouletteError::CallerIsNotTheHouseOwner)
+            }
+        }
+
+        // --- DANGER ZONE --
+
+        /// Withdraws the casino balance
+        ///
+        /// Can only be called by the house
+        #[ink(message)]
+        pub fn withdraw(&mut self, amount: Balance) -> Result<()> {
+            let caller = self.env().caller();
+            self.ensure_house(caller)?;
+            self.env().transfer(caller, amount)?;
+            Ok(())
+        }
+
+        /// Upgrades contract code
+        ///
+        /// Can only be called by the house
+        #[ink(message)]
+        pub fn set_code(&mut self, code_hash: [u8; 32], callback: Option<Selector>) -> Result<()> {
+            self.ensure_house(self.env().caller())?;
+            set_code_hash(&code_hash)?;
```

Comment:
Oh wow. I didn't know this was possible. Just like a runtime upgrade. So when one is considering using a real-world contract with real-world money, this is one of the things they should look for in the contract code.